source/Halibut/ServiceModel/PendingRequestQueueAsync.cs

```diff
@@ -54,50 +54,97 @@ public async Task<ResponseMessage> QueueAndWaitAsync(RequestMessage request, Req
             }
             finally
             {
-                using (await queueLock.LockAsync())
+                using (await queueLock.LockAsync(CancellationToken.None))
                 {
+                    queue.Remove(pending);
                     inProgress.Remove(request.Id);
                 }
             }
 
             return pending.Response;
         }
 
-        public bool IsEmpty => queue.IsEmpty;
-        public int Count => queue.Count;
+        public bool IsEmpty
+        {
+            get
+            {
+                using (queueLock.Lock(CancellationToken.None))
+                {
+                    return queue.Count == 0;
+                }
+
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                using (queueLock.Lock(CancellationToken.None))
+                {
+                    return queue.Count;
+                }
+
+            }
+        }
 
         public async Task<RequestMessage> DequeueAsync(CancellationToken cancellationToken)
         {
-            var pending = await DequeueNextAsync(cancellationToken);
-            if (pending == null) return null;
+            var timer = Stopwatch.StartNew();
```

Comment:
I wonder if this loop is having the desired outcome.

The aim is to get rid of queue items that are invalid.

But here, we start `DequeueNextAsync`, which can delay for the rest of the timeout. If that happens, and we find `BeginTransfer` returns a dud request, when we go around again, `DequeueNextAsync` will return null (as we reached the end of our time).

I wonder if instead of having a loop around the whole thing, we instead put the loop around `BeginTransfer`.

I.e., if we get something to transfer, and it becomes a dud, then keep pulling from the queue until it is either null or not a dud.

We don't really need to worry about the delay then, because the delay is really just to hang around until something appears on the list. This just becomes a tight "cleanup" loop. Plus it will cope when there are multiple duds on the queue.

Reply:
I've revisited this. Can you 👀 @sburmanoctopus